src/modules/renderPage.js

```diff
@@ -1,67 +1,103 @@
+		return;
+	}
+
+	const skipTransition = fragment || (popstate && !this.options.animateHistoryBrowsing);
 
 	// replace state in case the url was redirected
-	const url = new Link(page.responseURL).getPath();
+	const url = new Link(page.responseURL).getAddress();
 	if (window.location.pathname !== url) {
-		window.history.replaceState(
-			{
-				url,
-				random: Math.random(),
-				source: 'swup'
-			},
-			document.title,
-			url
-		);
+		updateHistoryRecord(url);
 
 		// save new record for redirected url
 		this.cache.cacheUrl({ ...page, url });
 	}
 
-	// only add for non-popstate transitions
-	if (!popstate || this.options.animateHistoryBrowsing) {
+	// only add for page loads with transitions
+	if (!skipTransition) {
 		document.documentElement.classList.add('is-rendering');
 	}
 
-	this.triggerEvent('willReplaceContent', popstate);
-	// replace blocks
-	for (let i = 0; i < page.blocks.length; i++) {
-		document.body.querySelector(`[data-swup="${i}"]`).outerHTML = page.blocks[i];
+	const replaceBlocks = () => {
+		this.triggerEvent('willReplaceContent', popstate);
+		page.blocks.forEach((html, i) => {
+			const block = query(`[data-swup="${i}"]`, document.body);
+			block.outerHTML = html;
+		});
+		this.triggerEvent('contentReplaced', popstate);
+	};
+
+	const replaceFragments = () => {
+		const fragments = Object.entries(page.fragments);
+		if (!fragments.length) {
+			console.warn('[swup] No fragments found, replacing whole page');
+			return false;
+		}
+
+		const { fragmentContainerAttr } = this.options;
+		const fragmentsOnPage = queryAll(`[${fragmentContainerAttr}]`).filter(el => fragment === true || fragment === el.getAttribute(fragmentContainerAttr));
+		const fragmentsToReplace = fragments.filter(([name]) => fragment === true || fragment === name);
+		const hasIdenticalFragmentContainers = compareArrays(
+			fragmentsOnPage.map((el) => el.getAttribute(fragmentContainerAttr)),
+			fragmentsToReplace.map(([name]) => name)
+		);
+
+		if (!hasIdenticalFragmentContainers) {
+			console.warn('[swup] Mismatching fragments on current and new page, replacing whole page');
+			return false;
+		}
+
+		if (typeof fragment === 'string' && !page.fragments[fragment]) {
+			console.warn(`[swup] Fragment "${fragment}" not found, replacing whole page`);
+			return false;
+		}
+
+		document.documentElement.classList.add('is-fragment');
+
+		this.triggerEvent('willReplaceContent', popstate);
+		this.triggerEvent('willReplaceFragments', popstate);
+		fragmentsToReplace.forEach(([name, html]) => {
+			const container = query(`[${fragmentContainerAttr}="${name}"]`);
+			if (container) {
+				container.outerHTML = html;
+			}
+		});
+		this.triggerEvent('contentReplaced', popstate);
+		this.triggerEvent('fragmentsReplaced', popstate);
+	};
+
+	// Try fragments first, if not, then blocks
+	if (fragment) {
+		if (replaceFragments() === false) {
+			console.log('Fragments failed, replacing blocks');
```

Comment:
I think better to keep the terminology consistent.

```suggestion
			console.log('Fragments failed, replacing containers');
```

Reply:
Sounds good!